File: Uber/204.count-primes.go
package leetcode

// CountPrimes204 is the Sieve of Eratosthenes. When p is prime, mark multiples
// from p*p because smaller multiples were marked by smaller primes.
//
// Time: O(n log log n). Space: O(n).
func CountPrimes204(n int) int {
	if n <= 2 {
		return 0
	}
	prime := make([]bool, n)
	for i := 2; i < n; i++ {
		prime[i] = true
	}
	for p := 2; p*p < n; p++ {
		if prime[p] {
			for m := p * p; m < n; m += p {
				prime[m] = false
			}
		}
	}
	count := 0
	for _, ok := range prime {
		if ok {
			count++
		}
	}
	return count
}

File: companies/Nokia/204.count-primes.cpp
#include <algorithm>
#include <array>
#include <cctype>
#include <climits>
#include <cstdlib>
#include <functional>
#include <numeric>
#include <queue>
#include <sstream>
#include <string>
#include <unordered_map>
#include <unordered_set>
#include <utility>
#include <vector>
using namespace std;


class Solution {
public:
    int countPrimes(int n) {
        /*
        Approach: Sieve of Eratosthenes. Mark multiples of each prime starting at
        p*p because smaller multiples were already marked by smaller primes.

        Complexity: O(n log log n) time, O(n) space.
        */
        if (n <= 2) return 0;
        vector<bool> prime(n, true);
        prime[0] = prime[1] = false;
        for (long long p = 2; p * p < n; ++p) {
            if (!prime[p]) continue;
            for (long long multiple = p * p; multiple < n; multiple += p) prime[multiple] = false;
        }
        return count(prime.begin(), prime.end(), true);
    }
};
